test/validation/api/ipsec/ipsec_test_out.c

```diff
@@ -1591,6 +1591,150 @@ static void test_test_sa_update_seq_num(void)
+#define HARD_LIMIT_PKT_CNT 2048
+#define DELTA_PKT_CNT 320
+
+static void test_out_ipv4_esp_sa_expiry(enum ipsec_test_sa_expiry expiry)
+{
+	int test_packet_len = pkt_ipv4_icmp_0.len - pkt_ipv4_icmp_0.l3_offset;
+	uint32_t src = IPV4ADDR(10, 0, 11, 2);
+	uint32_t dst = IPV4ADDR(10, 0, 22, 2);
+	odp_ipsec_tunnel_param_t out_tunnel;
+	odp_ipsec_sa_param_t param_out;
+	odp_ipsec_capability_t capa;
+	int i, inc, limit, delta;
+	uint64_t soft_limit_byte;
+	uint64_t hard_limit_byte;
+	uint64_t soft_limit_pkt;
+	uint64_t hard_limit_pkt;
+	odp_ipsec_sa_t out_sa;
+
+	switch (expiry)	{
+	case IPSEC_TEST_EXPIRY_SOFT_PKT:
+		soft_limit_pkt = SOFT_LIMIT_PKT_CNT;
+		hard_limit_pkt = HARD_LIMIT_PKT_CNT;
+		soft_limit_byte = 0;
+		hard_limit_byte = 0;
+		delta = DELTA_PKT_CNT;
+		limit = soft_limit_pkt;
+		inc = 1;
+		break;
+	case IPSEC_TEST_EXPIRY_HARD_PKT:
+		soft_limit_pkt = SOFT_LIMIT_PKT_CNT;
+		hard_limit_pkt = HARD_LIMIT_PKT_CNT;
+		soft_limit_byte = 0;
+		hard_limit_byte = 0;
+		delta = DELTA_PKT_CNT;
+		limit = hard_limit_pkt;
+		inc = 1;
+		break;
+	case IPSEC_TEST_EXPIRY_SOFT_BYTE:
+		soft_limit_pkt = 0;
+		hard_limit_pkt = 0;
+		soft_limit_byte = test_packet_len * SOFT_LIMIT_PKT_CNT;
+		hard_limit_byte = test_packet_len * HARD_LIMIT_PKT_CNT;
+		delta = test_packet_len * DELTA_PKT_CNT;
+		limit = soft_limit_byte;
+		inc = test_packet_len;
+		break;
+	case IPSEC_TEST_EXPIRY_HARD_BYTE:
+		soft_limit_pkt = 0;
+		hard_limit_pkt = 0;
+		soft_limit_byte = test_packet_len * SOFT_LIMIT_PKT_CNT;
+		hard_limit_byte = test_packet_len * HARD_LIMIT_PKT_CNT;
+		delta = test_packet_len * DELTA_PKT_CNT;
+		limit = hard_limit_byte;
+		inc = test_packet_len;
+		break;
+	default:
+		return;
+	}
+
+	memset(&out_tunnel, 0, sizeof(odp_ipsec_tunnel_param_t));
+
+	out_tunnel.type = ODP_IPSEC_TUNNEL_IPV4;
+	out_tunnel.ipv4.src_addr = &src;
+	out_tunnel.ipv4.dst_addr = &dst;
+
+	CU_ASSERT(odp_ipsec_capability(&capa) == 0);
+
+	ipsec_sa_param_fill(&param_out, ODP_IPSEC_DIR_OUTBOUND, ODP_IPSEC_ESP,
+			    0x4a2cbfe7, &out_tunnel,
+			    ODP_CIPHER_ALG_AES_CBC, &key_a5_128,
+			    ODP_AUTH_ALG_SHA1_HMAC, &key_5a_160,
+			    NULL, NULL);
+
+	param_out.lifetime.soft_limit.bytes = soft_limit_byte;
+	param_out.lifetime.hard_limit.bytes = hard_limit_byte;
+	param_out.lifetime.soft_limit.packets = soft_limit_pkt;
+	param_out.lifetime.hard_limit.packets = hard_limit_pkt;
+
+	out_sa = odp_ipsec_sa_create(&param_out);
+	CU_ASSERT_NOT_EQUAL_FATAL(ODP_IPSEC_SA_INVALID, out_sa);
+
+	ipsec_test_part test_out = {
+		.pkt_in = &pkt_ipv4_icmp_0,
+		.num_pkt = 1,
+		.out = {
+			{ .status.warn.all = 0,
+			  .status.error.all = 0,
+			  .l3_type = ODP_PROTO_L3_TYPE_IPV4,
+			  .l4_type = ODP_PROTO_L4_TYPE_ESP,
+			},
+		},
+	};
+
+	if ((suite_context.outbound_op_mode == ODP_IPSEC_OP_MODE_INLINE) &&
+	    (expiry == IPSEC_TEST_EXPIRY_HARD_PKT ||
+	     expiry == IPSEC_TEST_EXPIRY_HARD_BYTE)) {
+		test_out.flags.test_sa_expiry = true;
+		test_out.out[0].sa_expiry = IPSEC_TEST_EXPIRY_NONE;
```

Comment:
Why is this special handling for hard limits in inline mode needed here? To make sure status events are received if they come even before limit-delta packets? This seems a bit hacky and out-of-place here.

Reply:
When testing for hard expiry, before reaching the hard limit, the first loop of test packets would cross the soft expiry limit. With this special handling, the test expects the soft expiry warning event to be received anytime during the first loop itself.

In other words, so that the status event is not stuck for ever in case of hard expiry tests, this special handling was added.

Reply:
I think this should be handled more "automatically" by the generic code. I think now the test case code and the generic code are not well separated regarding the expiry checking.

Reply:
Ack. Realized that we could make use of IGNORE here to make the handling "automatic". Let me know if the test case code now looks OK.